sockeye/score.py

```diff
@@ -0,0 +1,119 @@
+Simple Training CLI.
+"""
+import argparse
+import os
+import sys
+from contextlib import ExitStack
+from typing import Any, cast, Optional, Dict, List, Tuple
+
+import mxnet as mx
+
+from . import arguments
+from . import constants as C
+from . import data_io
+from . import inference
+from . import model
+from . import scoring
+from . import train
+from . import utils
+from .log import setup_main_logger
+from .output_handler import get_output_handler, OutputHandler
+from .utils import check_condition, log_basic_info
+
+# Temporary logger, the real one (logging to a file probably, will be created in the main function)
+logger = setup_main_logger(__name__, file_logging=False, console=True)
+
+
+def main():
+    params = arguments.ConfigArgumentParser(description='Score data with an existing model.')
+    arguments.add_score_cli_args(params)
+    args = params.parse_args()
+    score(args)
+
+
+def score(args: argparse.Namespace):
+    global logger
+    logger = setup_main_logger(__name__, file_logging=False)
+
+    utils.log_basic_info(args)
+
+    with ExitStack() as exit_stack:
+        context = utils.determine_context(device_ids=args.device_ids,
+                                          use_cpu=args.use_cpu,
+                                          disable_device_locking=args.disable_device_locking,
+                                          lock_dir=args.lock_dir,
+                                          exit_stack=exit_stack)
+        if args.batch_type == C.BATCH_TYPE_SENTENCE:
+            check_condition(args.batch_size % len(context) == 0, "When using multiple devices the batch size must be "
+                                                                 "divisible by the number of devices. Choose a batch "
+                                                                 "size that is a multiple of %d." % len(context))
+        logger.info("Scoring Device(s): %s", ", ".join(str(c) for c in context))
+
+        model_config = model.SockeyeModel.load_config(os.path.join(args.model, C.CONFIG_NAME))
+
+        if args.max_seq_len is None:
+            max_seq_len_source = model_config.config_data.max_seq_len_source
+            max_seq_len_target = model_config.config_data.max_seq_len_target
+        else:
+            max_seq_len_source, max_seq_len_target = args.max_seq_len
+
+        # This call has a number of different parameters compared to training which reflect our need to get scores
+        # one-for-one and in order with the input data.
+        # Bucketing and permuting need to be turned off in order to preserve the ordering of sentences.
+        # The 'zeros' fill_up strategy fills underfilled buckets with zeros which can then be used to find the last item.
+        # Finally, 'resume_training' needs to be set to True because it causes the model to be loaded instead of initialized.
+        score_iter, _, config_data, source_vocabs, target_vocab, data_info = train.create_data_iters_and_vocabs(
+            args=args,
```

Comment:
Ok, got your point now, thanks for the explanation. I agree that passing args is somehow lazy, but one could say that your approach of reusing `create_data_iters_and_vocabs()` for scoring also is: you are going quite some way to overload the semantics of certain arguments, such that this function fits your needs (i.e. by enabling resume_training for something that is definitely not training), which effectively renders half of the (admittedly huge function) useless.
To me it looks like there are two paths forward to this: we either stay lazy, but then changing the args namespace would work equally well, would entail fewer changes in existing code, and is local to score.py; or we do a proper refactor of this by splitting up `create_data_iters_and_vocabs()` so that pieces that are needed by train.py and score.py can be reused.
I'd be fine either way :)

Reply:
This makes sense; I hadn't thought about the semantics of `resume_training`, for some reason. I agree with you now that we should either use `args` or a separate function. I'll look at this tomorrow and see which of your two options makes more sense.